allensdk/brain_observatory/ecephys/write_nwb/__main__.py

```diff
@@ -397,41 +397,94 @@ def add_probe_to_nwbfile(nwbfile, probe_id, sampling_rate, lfp_sampling_rate,
-        raise ValueError(
-            f"unable to recognize ids in this channel table. index: {channel_table.index}, columns: {channel_table.columns}"
-        )
+def add_ecephys_electrodes(nwbfile: pynwb.NWBFile,
+                           channels: List[dict],
+                           electrode_group: EcephysElectrodeGroup,
+                           local_index_whitelist: Optional[np.ndarray] = None):
+    """Add electrode information to an ecephys nwbfile electrode table.
 
-    channel_table["group"] = electrode_group
-    return channel_table
+    Parameters
+    ----------
+    nwbfile : pynwb.NWBFile
+        The nwbfile to add electrodes data to
+    channels : List[dict]
+        A list of 'channel' dictionaries containing the following fields:
+            id: The unique id for a given electrode/channel
+            probe_id: The unique id for an electrode's/channel's device
+            valid_data: Whether the data for an electrode/channel is usable
+            local_index: The local index of an electrode/channel on a given device
+            probe_vertical_position: Length-wise position of electrode/channel on device (microns)
+            probe_horizontal_position: Width-wise position of electrode/channel on device (microns)
+            manual_structure_id: The LIMS id associated with an anatomical structure
+            manual_structure_acronym: Acronym associated with an anatomical structure
+            anterior_posterior_ccf_coordinate
+            dorsal_ventral_ccf_coordinate
+            left_right_ccf_coordinate
+    electrode_group : EcephysElectrodeGroup
+        The pynwb electrode group that electrodes should be associated with
+    local_index_whitelist : Optional[np.ndarray], optional
+        If provided, only add electrodes (a.k.a. channels) specified by the
+        whitelist (and in order specified), by default None
+    """
+    add_ecephys_electrode_columns(nwbfile)
+
+    channel_table = pd.DataFrame(channels)
+
+    if local_index_whitelist is not None:
+        channel_table.set_index("local_index", inplace=True)
+        channel_table = channel_table.loc[local_index_whitelist, :]
+        channel_table.reset_index(inplace=True)
+
+    for _, row in channel_table.iterrows():
+        x = row["anterior_posterior_ccf_coordinate"]
+        y = row["dorsal_ventral_ccf_coordinate"]
+        z = row["left_right_ccf_coordinate"]
+
+        nwbfile.add_electrode(
+            id=row["id"],
+            x=(np.nan if x is None else x),  # Not all probes have CCF coords
+            y=(np.nan if y is None else y),
+            z=(np.nan if z is None else z),
+            probe_vertical_position=row["probe_vertical_position"],
+            probe_horizontal_position=row["probe_horizontal_position"],
+            local_index=row["local_index"],
+            valid_data=row["valid_data"],
+            probe_id=row["probe_id"],
+            group=electrode_group,
+            imp=150000.0,
```

Comment:
hardcoded?

Reply:
@jsiegle Is there a better value that I can use for the impedence of the neuropixels electrode? I got that value from the [neuropixels 1.0 brochure pdf](https://e2f49b05-3010-45e2-a7bf-b8f67811425a.filesusr.com/ugd/832f20_4a14406ba1204e60ae8534b09e201b49.pdf) but it's not likely to be accurate for physiological recordings...

Would you prefer I left it as a blank string? `""`. The `nwbfile.add_electrode` method requires something be recorded for `imp`.

Reply:
Leave it blank. We can't measure the true impedance for these electrodes.